rewriter/Data.cc

```diff
@@ -0,0 +1,136 @@
+#include "rewriter/Data.h"
+#include "absl/strings/match.h"
+#include "ast/Helpers.h"
+#include "ast/ast.h"
+#include "core/Context.h"
+#include "core/Names.h"
+#include "core/core.h"
+#include "core/errors/rewriter.h"
+#include "rewriter/Util.h"
+#include "rewriter/rewriter.h"
+
+using namespace std;
+
+namespace sorbet::rewriter {
+
+namespace {
+
+bool isMissingInitialize(const core::GlobalState &gs, const ast::Send *send) {
+    if (!send->hasBlock()) {
+        return true;
+    }
+
+    auto block = send->block();
+
+    if (auto *insSeq = ast::cast_tree<ast::InsSeq>(block->body)) {
+        auto methodDef = ast::cast_tree<ast::MethodDef>(insSeq->expr);
+
+        if (methodDef && methodDef->name == core::Names::initialize()) {
+            return false;
+        }
+
+        for (auto &&stat : insSeq->stats) {
+            methodDef = ast::cast_tree<ast::MethodDef>(stat);
+
+            if (methodDef && methodDef->name == core::Names::initialize()) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+} // namespace
+
+vector<ast::ExpressionPtr> Data::run(core::MutableContext ctx, ast::Assign *asgn) {
+    vector<ast::ExpressionPtr> empty;
+
+    if (ctx.state.runningUnderAutogen) {
+        return empty;
+    }
+
+    auto lhs = ast::cast_tree<ast::UnresolvedConstantLit>(asgn->lhs);
+    if (lhs == nullptr) {
+        return empty;
+    }
+
+    auto send = ast::cast_tree<ast::Send>(asgn->rhs);
+    if (send == nullptr) {
+        return empty;
+    }
+
+    auto recv = ast::cast_tree<ast::UnresolvedConstantLit>(send->recv);
+    if (recv == nullptr) {
+        return empty;
+    }
+
+    if (!ast::MK::isRootScope(recv->scope) || recv->cnst != core::Symbols::Data().data(ctx)->name ||
+        send->fun != core::Names::define() || !send->hasPosArgs()) {
+        return empty;
+    }
+
+    auto loc = asgn->loc;
+
+    ast::MethodDef::ARGS_store newArgs;
+    ast::Send::ARGS_store sigArgs;
+    ast::ClassDef::RHS_store body;
+
+    for (int i = 0; i < send->numPosArgs(); i++) {
+        auto *sym = ast::cast_tree<ast::Literal>(send->getPosArg(i));
+        if (!sym || (!sym->isSymbol() && !sym->isString())) {
+            return empty;
+        }
+        core::NameRef name = sym->asSymbol();
+        auto symLoc = sym->loc;
+        auto strname = name.shortName(ctx);
+        if (!strname.empty() && strname.back() == '=') {
+            if (auto e = ctx.beginError(symLoc, core::errors::Rewriter::InvalidStructMember)) {
+                e.setHeader("Data member `{}` cannot end with an equal", strname);
+            }
+        }
+
+        // TODO(jez) Use Loc::adjust here
```

Comment:
This file started out as a copy+paste of the [Struct rewriter](https://github.com/sorbet/sorbet/blob/master/rewriter/Struct.cc#L129), which has a few comments that we kept in the file. I'm happy to remove them if folks would prefer.